watertap/unit_models/crystallizer.py

```diff
@@ -0,0 +1,696 @@
+        def eq_dens_magma(b):
+            return (self.magma_density == 
+                b.properties_solids[0].dens_mass_solute['Sol'] * self.product_volumetric_solids_fraction
+                )
+
+        # (d) Operating pressure constraint
+        @self.Constraint(doc="Operating pressure constraint")
+        def eq_operating_pressure_constraint(b):
+            return (self.pressure_operating - b.properties_out[0].pressure_sat == 0)
+
+        # (e) Relative supersaturation 
+        @self.Constraint(solute_set, doc="Relative supersaturation created via evaporation, g/g (solution)")
+        def eq_relative_supersaturation(b, j):
+            #  mass_frac_after_evap = SOLIDS IN + LIQUID IN - VAPOUR OUT
+            mass_frac_after_evap = b.properties_in[0].flow_mass_phase_comp['Liq', j] \
+            / (sum(b.properties_in[0].flow_mass_phase_comp['Liq', j] for j in solute_set)
+                + b.properties_in[0].flow_mass_phase_comp['Liq', 'H2O']
+                - b.properties_vapor[0].flow_mass_phase_comp['Vap', 'H2O']
+                )
+            # return (b.relative_supersaturation[j] * b.properties_out[0].solubility_mass_frac_phase_comp['Liq', j] == 
+            # (mass_frac_after_evap - b.properties_out[0].solubility_mass_frac_phase_comp['Liq', j])
+            # )
+            return (b.relative_supersaturation[j] == 
+                (mass_frac_after_evap - b.properties_out[0].solubility_mass_frac_phase_comp['Liq', j])
+                / b.properties_out[0].solubility_mass_frac_phase_comp['Liq', j]
+                )
+
+
+
+
+
+
+        # 4. Fix flows of empty solid, liquid and vapour streams
+        # (i) Fix solids: liquid and vapour flows must be zero
+        @self.Constraint(self.properties_solids[0].phase_component_set,
+            doc="Empty components in solids stream")
+        def eq_solids_composition(b, p, j):
+            if p != 'Sol':
+                return (b.properties_solids[0].flow_mass_phase_comp[p, j] == 1e-8 * pyunits.kg/pyunits.s)
+            else:
+                return Constraint.Skip
+
+        # (ii) Fix liquids: solid and vapour flows must be zero
+        @self.Constraint(self.properties_out[0].phase_component_set,
+            doc="Empty components in liquid stream")
+        def eq_liquid_composition(b, p, j):
+            if p != 'Liq':
+                return (b.properties_out[0].flow_mass_phase_comp[p, j] == 1e-8 * pyunits.kg/pyunits.s)
+            else:
+                return Constraint.Skip
+
+        # (iii) Fix vapor: solid and vapour flows must be zero. Vapour is also pure solvent.
+        @self.Constraint(self.properties_vapor[0].phase_component_set,
+            doc="Empty components in vapour stream")
+        def eq_vapor_composition(b, p, j):
+            if p != 'Vap':
+                return (b.properties_vapor[0].flow_mass_phase_comp[p, j] == 1e-8 * pyunits.kg/pyunits.s)
+            else:
+                return Constraint.Skip
```

Comment:
Instead of setting the variables equal to a constant via a constraint, it would be better to just fix the variables.

Reply:
@TimBartholomew  So to be clear, you'd rather this was written as:

```
        for p, j in self.properties_solids[0].phase_component_set:
            if p != 'Sol':
                self.properties_solids[0].flow_mass_phase_comp[p, j].fix(1e-8)

        for p, j in self.properties_out[0].phase_component_set:
            if p != 'Liq':
                self.properties_out[0].flow_mass_phase_comp[p, j].fix(1e-8)

        for p, j in self.properties_vapor[0].phase_component_set:         
            if p != 'Vap':
                self.properties_vapor[0].flow_mass_phase_comp[p, j].fix(1e-8)
```

Reply:
Done.